### DISCUSSION TITLE
Build own mode Night-Lapse Photo

### DISCUSSION THREAD

**Discussion by tomas123** [gopro/labs#511] [Show and tell]
Problem of Night-Lapse Photo Mode
`While sundown we need the automatic mode for iso and shutter speed.
After sunset, the Hero 11's auto exposure works very poorly. The starry sky is much too dark.`

Left: Auto-shutter / Auto-ISO(crop from jpg)
Right: 30s/ISO800  (crop from jpg)
![](https://user-images.githubusercontent.com/222947/263552002-9c5cdc6d-cb3e-4ae9-a0ea-432afc022a5a.jpeg)

I wrote here same working code for injection of new settings while Night-Lapse Photo Mode is running:
https://github.com/gopro/labs/discussions/500
The problem is that the EV metering (i isoValue / s shutterValue) is independent of the actual exposure time in nightlapse mode.

-----

That's why I switched to programming the code for the loop "mNP - mode Night-Lapse Photo" myself.

I use two timers. 
A timer measures the time of the main loop for "mPN - mode Photo Night" and a second timer measures how long the status "r recording" is active.

main loop is 30s
- while sunset: Auto-ISO and Auto-Shutter
- in the night: ISO800 and 30s

`<l1t+i8M1+mPN+eA+r!S=At:B=Bt:B>r0!R25=B-A>B10t+i8M8+e30=Ct:B=C-A"$C\n$B"<C30!R55!R`

pseudo code
````
<l1    //if new loop starts set Sundown-Mode 
     t+i8M1      //ISO 100-800
     +mPN        //mode Photo Night
     +eA         //Auto Shutter
     +r          //RAW + JPG
!S        //start
=At:B   //A = t:X timedate with B-seconds_since_Boot
// do 
     =Bt:B    //B = seconds_since_Boot
     >r0!R25    
// while  recording >= 0

=B-A     //B=B-A as equivalent to exposure time  
>B10    // if "equivalent to exposure time" > 10 seconds
    t+i8M8 // fix ISO 800
    +e30    // Shutter 30s
//do 
    =Ct:B  //B = seconds_since_Boot
    =C-A  //C = loop timer 
    "$C\n$B" // print loop timer and "equivalent to exposure time"
    <C30!R55  
// while Loop Timer < 30 seconds

!R    //repeat main loop 
````

With this timer I can measure the actual exposure time with variable B very well:
exposure time of taken photo 
= 1/1000 -> B=1.6s
= 8s -> B=10s
= 10s -> B=12.5s

so i chose B=10 in the loop and it works great
````
======== /Volumes/256GB/DCIM/107GOPRO/GOPR4130.JPG
ISO                             : 800
Shutter Speed                   : 7
======== /Volumes/256GB/DCIM/107GOPRO/GOPR4131.JPG
ISO                             : 800
Shutter Speed                   : 8    // last Photo with Autoshutter 
======== /Volumes/256GB/DCIM/107GOPRO/GOPR4132.JPG
ISO                             : 800
Shutter Speed                   : 30  // Shutter 30s 
````

Left: ISO 800, Last Image with Auto Shutter (8s)
Right the next Image: ISO 800, 30s Shutter
![Aggregated](https://github.com/gopro/labs/assets/222947/4c79b502-9354-4635-a926-3a5679408e1f)
I adjusted here the brightness of the images in lightroom to create a smooth image transition.
In my workflow I use LRTimelapse to combine the RAW images into a video and compensate jumps in exposure.

**Comment by tomas123** [comment]
it's a bit tricky but Blockly can do it too
https://gopro.github.io/labs/build/
A big advantage is avoiding byte counting for jumps!!

**//UPDATE//**
This version automatically switches from sunset (auto shutter) to starry sky (fixed shutter) and sunrise (auto shutter)
For details for EV metering (i isoValue / s shutterValue) see my post here 
https://github.com/gopro/labs/discussions/500

You can readout the own defined EV metadata tag with exiftool
```
% exiftool -iso -shutterspeed -createdate -u -gopro_eval GOPR*.JPG 
======== GOPR4733.JPG
ISO                             : 800
Shutter Speed                   : 30
Create Date                     : 2023:09:02 02:05:33
GoPro EVAL                      : 0.0495436489582062
```
function:
- main loop is 30s (+2s overlay)
- while sunset and sunrise: Night-Lapse Photo with Auto-ISO and Auto-Shutter
- in the night: Night-Lapse Photo with ISO800 and 30s
- switch from sunset Auto-Shutter to Night by exposure time 10s (auto shutter)
- switch from night (with fixed exposure time 30s) to sunrise (auto-shutter) at EV 0.1

code
`ti8M1mPNeAr=F1!S=At:B=Bt:B=B-A>r0=CB+!R21=E6.25=Gi=G*0.01=E/G=E*s=E#2oMEVAL=E"$Bs - $Cs - EV$E">z1!X<B30!R21<F1>E0.1!R>C10t+i8M8+e30+=F0!R14
`

blockly
variable B: loop timer 30s
variable C: auto exposure timer 0..10s
variable E: EV value from metering system (iso and shutter)
variable F: status Flag 0/1 = night/twilight 
<img width="480" alt="Bildschirmfoto 2023-09-02 um 16 14 18" src="https://github.com/gopro/labs/assets/222947/ab963b3b-2709-49a5-803b-2820ba9e3535">

background: 
The deepest values for i (isoValue) and s (shutterValue) are  1/2.6087s and  ISO1575 -> EV +0.0495
That's why you can only switch back to auto exposure at EV>0.0495. 
Unfortunately, this is a bit to late at sunrise ... (same frames are overexposed at 30s)

![New Project-9_2_2023, 16_13_39](https://github.com/gopro/labs/assets/222947/8199b81e-3fa4-428e-92d4-51880a6ff499)

----

a note for interested users - you can download the QR code and read it in Blockly!
<img width="118" alt="Bildschirmfoto 2023-09-01 um 11 53 16" src="https://github.com/gopro/labs/assets/222947/26619351-2584-4ee0-acd9-c1c22bd66cf3">
the block code is stored as xml in the png header 

PS: I miss loop3 in Blockly ;-)

**Reply by dnewman-gpsw** [reply]
Loop3 can be added, as with loop4, etc.

**Comment by tomas123** [comment]
but how?
<img width="154" alt="Bildschirmfoto 2023-09-02 um 18 42 39" src="https://github.com/gopro/labs/assets/222947/5266e260-cb63-43ec-ae06-e19e05704134">

**Reply by dnewman-gpsw** [reply]
Sorry, I meant the tool can be expanded with coding.  Is loop3 the only thing you are missing?

**Reply by tomas123** [reply]
yeah sure, 
counting the bytes for jumps is hell

**Reply by dnewman-gpsw** [reply]
Loop3 and Loop4 has now be implemented and is live.

**Reply by tomas123** [reply]
thank you

**Reply by woksio** [reply]
hey, i'd like to use your approach to shoot sunset, then stars then sunrise at one go with gopro changing shutter etc.

i used your and it seem it starts but going to max 10s for shutter so going for starts would be too short.
did you anyhow modify the code maybe?
